model/access.go

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2015 Spinpunch, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package model
+
+import (
+	"encoding/json"
+	"io"
+)
+
+const (
+	ACCESS_TOKEN_EXPIRE_TIME = -1 // forever
+	ACCESS_TOKEN_CACHE_SIZE  = 10000
+	ACCESS_TOKEN_GRANT_TYPE  = "authorization_code"
+	ACCESS_TOKEN_TYPE        = "bearer"
+	REFRESH_TOKEN_GRANT_TYPE = "refresh_token"
+)
+
+type AccessData struct {
+	AuthCode     string `json:"auth_code"`
+	UserId       string `json:"user_id"`
+	Token        string `json:"token"`
+	RefreshToken string `json:"refresh_token"`
+	ExpiresIn    int32  `json:"expires_in"`
+	CreateAt     int64  `json:"create_at"`
+	RedirectUri  string `json:"redirect_uri"`
+	Scope        string `json:"scope"`
+}
+
+type AccessResponse struct {
+	AccessToken  string `json:"access_token"`
+	TokenType    string `json:"token_type"`
+	ExpiresIn    int32  `json:"expires_in"`
+	Scope        string `json:"scope"`
+	RefreshToken string `json:"refresh_token"`
+}
+
+// IsValid validates the AccessData and returns an error if it isn't configured
+// correctly.
+func (ad *AccessData) IsValid() *AppError {
+
+	if len(ad.AuthCode) == 0 || len(ad.AuthCode) > 128 {
+		return NewAppError("AccessData.IsValid", "Invalid auth code", "")
+	}
+
+	if len(ad.UserId) != 26 {
+		return NewAppError("AccessData.IsValid", "Invalid user id", "")
+	}
+
+	if len(ad.Token) == 0 || len(ad.Token) > 128 {
+		return NewAppError("AccessData.IsValid", "Invalid access token", "")
+	}
+
+	if len(ad.RefreshToken) > 128 {
+		return NewAppError("AccessData.IsValid", "Invalid refresh token", "")
+	}
+
+	if ad.ExpiresIn == 0 {
+		return NewAppError("AccessData.IsValid", "Expires in must be set", "")
+	}
+
+	if ad.CreateAt == 0 {
+		return NewAppError("AccessData.IsValid", "Create at must be a valid time", "")
+	}
+
+	if len(ad.RedirectUri) > 256 {
+		return NewAppError("AccessData.IsValid", "Invalid redirect uri", "")
+	}
+
+	if len(ad.Scope) > 128 {
+		return NewAppError("AccessData.IsValid", "Invalid scope", "")
+	}
+
+	return nil
+}
+
+func (ad *AccessData) PreSave(aesKey string) *AppError {
+	if ad.ExpiresIn == 0 {
+		ad.ExpiresIn = ACCESS_TOKEN_EXPIRE_TIME
+	}
+
+	if ad.CreateAt == 0 {
+		ad.CreateAt = GetMillis()
+	}
+
+	var err *AppError
+	if len(ad.Token) > 0 {
+		ad.Token, err = AesEncrypt(aesKey, ad.Token)
+	}
+
+	if len(ad.RefreshToken) > 0 {
+		ad.RefreshToken, err = AesEncrypt(aesKey, ad.RefreshToken)
+	}
+
+	return err
+}
+
+func (ad *AccessData) ToJson() string {
+	b, err := json.Marshal(ad)
+	if err != nil {
+		return ""
+	} else {
+		return string(b)
+	}
+}
+
+func AccessDataFromJson(data io.Reader) *AccessData {
+	decoder := json.NewDecoder(data)
+	var ad AccessData
+	err := decoder.Decode(&ad)
+	if err == nil {
+		return &ad
+	} else {
+		return nil
+	}
+}
+
+func (ad *AccessData) IsExpired() bool {
+
+	if ad.ExpiresIn < 0 {
```

Comment:
It is not, because you are currently using negatives as infinite. Which should be temporary.